components/ai_chat/core/common/prefs.cc

```diff
@@ -149,4 +241,90 @@ std::optional<base::Value::Dict> GetUserMemoryDictFromPrefs(
   return user_memory;
 }
 
+std::vector<mojom::SmartModePtr> GetSmartModesFromPrefs(
+    const PrefService& prefs) {
+  const base::Value::Dict& smart_modes_dict =
+      prefs.GetDict(prefs::kBraveAIChatSmartModes);
+
+  std::vector<mojom::SmartModePtr> smart_modes;
+
+  for (const auto [id, mode_value] : smart_modes_dict) {
+    if (!mode_value.is_dict()) {
+      continue;
+    }
+
+    auto smart_mode = SmartModeDictToStruct(id, mode_value.GetDict());
+    if (smart_mode) {
+      smart_modes.push_back(std::move(smart_mode));
+    }
+  }
+
+  return smart_modes;
+}
+
+mojom::SmartModePtr GetSmartModeFromPrefs(const PrefService& prefs,
+                                          const std::string& id) {
+  const base::Value::Dict& smart_modes_dict =
+      prefs.GetDict(prefs::kBraveAIChatSmartModes);
+
+  const base::Value* mode_value = smart_modes_dict.Find(id);
+  if (!mode_value || !mode_value->is_dict()) {
+    return nullptr;
+  }
+
+  return SmartModeDictToStruct(id, mode_value->GetDict());
+}
+
+void AddSmartModeToPrefs(const std::string& shortcut,
+                         const std::string& prompt,
+                         const std::optional<std::string>& model,
+                         PrefService& prefs) {
+  if (prompt.empty() || !IsValidAndUniqueShortcut(prefs, shortcut, "")) {
+    return;
+  }
```

Comment:
I saw we have maximum 1000 char length limit in spec, shouldn't we check it here? 
Also should we enforce shortcut length limit too?

Reply:
We do enforce 1000 limit in the frontend side to give a real-time feedback direct in frontend which would prevent user from saving it. Didn't have one for shortcut before, I'll add 64 for shortcut to the frontend.